.gitignore

```diff
@@ -5,3 +5,5 @@
 test/.vscode
 build
 .cache
+.idea
+**cmake-build**

```

Comment:
Does your IDE really require building inline with the sources, or within the source tree?
I am loathe to pollute the repository with support for the anti-pattern of building within the sources; however, Visual Studio Code forced me to yield.  VSCode is a poor tool that glitches in very frustrating ways when asked to build out of the source tree, that's the only reason for the `test/.vscode`, `build` and `.cache`.
So, unless your tool glitches, adopt the practice of not building in-tree and remove this.

Reply:
Why is this such a bad thing? I think most tools / LSPs expect to build in the source tree? I think CLion does this by default and I think `clangd` by default looks for `./build` in the root of the project?

Reply:
Just two things I dislike:
Most tools like the command line `grep`, `find`, etc. are extremely hard to configure for understanding `.gitignore`, so, you just slow down your work flow by having to treat the non-sources directories as explicit special cases with every non-git tool you use.

Let's say you want to build with GCC and also with Clang.
Who gets to build in-tree? both?
So, for every flavor of build you may want you need to provide an entry point within the tree, and add it to the .gitignore.
How's that helpful instead of segregating builds from sources?

These are two instances of the more general problem of disorganization.  A freshly cloned repository ought to contain only files generated by people or by tools that are so sophisticated they may as well.  Further processing of "sources" ought not to dump the results into a place where it is hard to tell them apart from files made by people.

It baffles me that colleagues expect that whatever configuration is good for holding source files is going to be just as good for built files.  For example, what if I have a configuration of very small blocks at the file system level expecting that source code is among the very small files in the distribution of file size, but keep the build filesystem with large blocks, because of the same reason on the opposite?

See? there are multiplicity of day to day advantages of keeping things organized.  Very fundamental stuff for Software Engineers.

Reply:
Sorry took a while to get back to this. 

I understand both of these points, I think the overriding point from me regardless of my opinion is that this is how it is going to be used by people, so we might as well make it convenient for people to contribute. Many IDEs and tools like the build folder to be in the source tree. I'm using the `clangd` LSP which is hooking into nvim which I believe requires the `compile_commands.json` to be in `./build`. JetBrains also will build in the source tree by default into `cmake-build-[BUILD_TYPE]`.  

Secondly, you can use tools like `ripgrep` in place of grep which are `.gitignore` sensitive, which is awesome.

Though I hear your points well and agree, this addition doesn't prevent more optimal workflows and enables anyone to contribute more easily.

Reply:
Let's continue this conversation online, at this point, it is an item that should not go in this PR, but in another PR if we decide to support building in-tree.

Let's have PRs with a "single responsibility".